source/mainnet/smart-contracts/tutorials/counter/counter-contract.rst

```diff
@@ -0,0 +1,230 @@
+        /// Your error
+        OwnerError,
+        IncrementError,
+        DecrementError,
+    }
+
+    /// Init function that creates a new smart contract.
+    #[init(contract = "counter")]
+    fn init<S: HasStateApi>(
+        _ctx: &impl HasInitContext,
+        _state_builder: &mut StateBuilder<S>,
+    ) -> InitResult<State> {
+        // Your code
+
+        Ok(State { counter: 0 })
+    }
+
+Increment and decrement counter
+===============================
+
+Increment counter
+-----------------
+
+Then change the update function as described below, let’s remember what we want; input needs to be parsed without any errors, the value should be positive(otherwise we will get an Error::IncrementError and the transaction should be triggered by the owner of the contract instance or it will throw Error::OwnerError and the function itself has to be a mutable function because we are going to change the state of the contract.
+
+.. code-block:: rust
+
+    type IncrementVal = i8;
+    /// Receive function. The input parameter is the boolean variable `throw_error`.
+    ///  If `throw_error == true`, the receive function will throw a custom error.
+    ///  If `throw_error == false`, the receive function executes successfully.
+    #[receive(
+        contract = "counter",
+        name = "increment",
+        parameter = "i8",
+        error = "Error",
+        mutable
+    )]
+    fn increment<S: HasStateApi>(
+        ctx: &impl HasReceiveContext,
+        host: &mut impl HasHost<State, StateApiType = S>,
+    ) -> Result<(), Error> {
+        // Your code
+
+        let param: IncrementVal = ctx.parameter_cursor().get()?;
+        let state = host.state_mut();
+        ensure!(
+            ctx.sender().matches_account(&ctx.owner()),
+            Error::OwnerError
+        );
+
+        ensure!(param > 0, Error::IncrementError);
+        state.counter += param;
+        Ok(())
+    }
+
+Decrement counter
+-----------------
+
+Add a new mutable function to implement decrement with a similar approach. It will also take an input parameter but this time we need to make sure that it is negative because a violation will be caused by an Error::DecrementError, like the other one this can be triggered by only the owner of the contract-otherwise it will throw an Error::OwnerError.
+
+.. code-block:: rust
+
+    #[receive(
+        contract = "counter",
+        name = "decrement",
+        parameter = "i8",
+        error = "Error",
+        mutable
+    )]
+    fn decrement<S: HasStateApi>(
+        ctx: &impl HasReceiveContext,
+        host: &mut impl HasHost<State, StateApiType = S>,
+    ) -> Result<(), Error> {
+        // Your code
+
+        let param: IncrementVal = ctx.parameter_cursor().get()?;
+        let state = host.state_mut();
+        ensure!(
+            ctx.sender().matches_account(&ctx.owner()),
+            Error::OwnerError
+        );
+
+        ensure!(param < 0, Error::DecrementError);
+        state.counter -= param;
```

Comment:
So you ensure it is negative, but then you subtract it. Is that desired? Calling it `decrement` in that case seems misleading.

Reply:
Line 132 should be "state.counter += param; "